### PR TITLE
Add data cleanse component

### PR DESCRIPTION
**Pull request by AdRiley**
### Pull Request Description

Add new cleanse and text_cleanse components

image

image

image

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: test/Table_Tests/src/In_Memory/Text_Cleanse_Spec.enso [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            expected_col3 = Column.from_vector "Test" ["String with Leading Spaces then Leading Numbers", "String with Leading Numbers then Leading Spaces"]
+            res1.should_equal expected_col1
+            res2.should_equal expected_col2
+            res3.should_equal expected_col3
+    suite_builder.group "Text Cleanse" group_builder->
+        input = "  It was the best  of  times  it was  the  worst  of  times  "
+        group_builder.specify "should remove leading whitespace" <|
+            expected = "It was the best  of  times  it was  the  worst  of  times  "
+            res = input.cleanse [Text_Cleanse.Leading_Whitespace]
+            res.should_equal expected
```
Comment (radeusgd) [inline test/Table_Tests/src/In_Memory/Text_Cleanse_Spec.enso]: It feels like this group belongs in `Base_Tests`, not here.

**Review by radeusgd — APPROVED** [review]
Looks good.

I'd prefer to move the `Text` type tests to `Base_Tests` as testing them as part of in-memory Table implementation feels weird.

**Comment by somebody1234** [comment]
would we perhaps want a `input.replace "\d" " "` cleanse too? (like Duplicate_Whitespace, except always replacing with spaces - for example, `'foo bar\t\t baz'`)

although it probably wouldn't be too late to only add that only when someone actually has a usecase for it...

**Comment by AdRiley** [comment]
> would we perhaps want a `input.replace "\d" " "` cleanse too? (like Duplicate_Whitespace, except always replacing with spaces - for example, `'foo bar\t\t baz'`)
> 
> although it probably wouldn't be too late to only add that only when someone actually has a usecase for it...

So operator78170.replace (regex "\d") ' ' exists today. But do you mean the ability to use our named regexs in replace?

That is an interesting idea...

**Comment by radeusgd** [comment]
> > would we perhaps want a `input.replace "\d" " "` cleanse too? (like Duplicate_Whitespace, except always replacing with spaces - for example, `'foo bar\t\t baz'`)
> > although it probably wouldn't be too late to only add that only when someone actually has a usecase for it...
> 
> So operator78170.replace (regex "\d") ' ' exists today. But do you mean the ability to use our named regexs in replace?
> 
> That is an interesting idea...

I understood as ability for the regex to replace the number not with empty `""`, but with single whitespace `" "`. Not sure if that was what you meant @somebody1234 ?

But I'm writing because this also struck a chord with me - I was thinking that with this method when cleaning e.g. `[a,b,c]` from all non-letters, I will get `abc`. Often that is what I want. But it feels to me that I may also want to get `a b c`. For example for language processing tasks, if I want to do some naive cleanup of punctuation before tokenization, I want `foo:bar, baz...     Hmm?` to probably become `foo bar baz Hmm`, so that I can then split it on `" "` to get all the tokens. (Of course if `foo:bar` should become `foo bar` or actually `foobar` is highly use-case dependent.)

But essentially this stems the idea if maybe we should be able to control if the cleansing should "preserve separation between words". I.e. by default we replace everything with `""`, but we could have an alternative mode where we kind of replace everything with `" "` and then do remove duplicate whitespace as the last step to normalize all separators to be a single space.

But it feels like this is complicating this rather simple tool, so maybe that is not really what we want at this stage for this component. Just throwing ideas around.

**Comment by somebody1234** [comment]
whoops my bad, i meant `"\s+" " "` :sweat_smile:

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 2]
```diff
@@ -531,6 +532,39 @@ Text.replace self term:(Text | Regex) replacement:Text (case_sensitivity:Case_Se
             updated_regex = term.recompile case_sensitivity
             updated_regex.replace self replacement only_first
 
+## GROUP Text
+   ICON text
+   Removes the specified cleansings from the text.
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Maybe it should be phrased as "Applies the specified cleansings to the text"?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 3]
```diff
@@ -531,6 +532,39 @@ Text.replace self term:(Text | Regex) replacement:Text (case_sensitivity:Case_Se
...
+
+   Arguments:
+   - remove: A vector of the text cleanings to remove from the text. The text cleansing are
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]:
```suggestion
   - remove: A vector of the text cleanings to remove from the text. The text cleansings are 
```

**Review by GregoryTravis — APPROVED** [review]

File: test/Table_Tests/src/In_Memory/Text_Cleanse_Spec.enso (hunk 1 above)
Comment (AdRiley) [reply test/Table_Tests/src/In_Memory/Text_Cleanse_Spec.enso]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 2 above)
Comment (AdRiley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 3 above)
Comment (AdRiley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Done

**Pull request merged**